V0.55.0 Release Notes (#4033)
grafana/k6: release notes/v0.55.0.md

k6 `v0.55.0` is here 🎉! This release includes:

> - ⚠️ The deprecated StatsD output has been removed.

Reviewer: Not sure if we need here emoji, keeping in mind that this already quite “urgent” section.

Author: I thought this added a sweet touch that attracted the eye to something users should not overlook. Feel free to remove it if you prefer without it 👍🏻

Author: Trying those out 🚀!

- ⚠️ The experimental `k6/experimental/tracing` module has been removed.
- URL grouping support in the browser module.
- Top-level `await` support.
- Complete RSA support for `k6/experimental/webcrypto`.

## Breaking changes

### `k6/experimental/tracing` module removed [`#3855`](https://github.com/grafana/k6/pull/3855)

> The experimental `k6/experimental/tracing` module has been removed, in favor of a replacement [jslib](https://jslib.k6.io/http-instrumentation-tempo/), please look at the [docs on how to transition](https://grafana.com/docs/k6/latest/javascript-api/jslib/http-instrumentation-tempo/#migration-from-k6experimentaltracing), [`#3855`](https://github.com/grafana/k6/pull/3855).

Reviewer 2: "https://jslib.k6.io/http-instrumentation-tempo/" -- That links points to the old k6.io site, and it's currently 404ing. Does the URL need to be updated?

Reviewer 2: ```suggestion
The experimental `k6/experimental/tracing` module has been removed, in favor of a replacement [jslib](https://jslib.k6.io/http-instrumentation-tempo/) polyfill, please look at the [docs on how to transition](https://grafana.com/docs/k6/latest/javascript-api/jslib/http-instrumentation-tempo/#migration-from-k6experimentaltracing), [`#3855`](https://github.com/grafana/k6/pull/3855).
```

Is it being replaced with a polyfill?

### StatsD output removed [`#3849`](https://github.com/grafana/k6/pull/3849)

The StatsD output was deprecated in k6 v0.47.0 and is now removed. You could still output results to StatsD using the community xk6 extension [LeonAdato/xk6-output-statsd](https://github.com/LeonAdato/xk6-output-statsd). Thanks, @LeonAdato for taking over the extension!

### `open` will have a breaking change in the future.

> Currently `open` opens relative files based on a strange root the same `require` before it was fixed around ESM. As part of actually making k6 consistent `open` among other functions like it will start to be relative in the same way as imports and `require` are.

Reviewer 2: ```suggestion
Currently, `open` opens relative files based on an unusual root, similar to how `require` behaved before it was updated for ESM compatibility. To make k6 more consistent, `open` and other functions like it will start handling relative paths in the same way as imports and `require`.
```

WDYT?

> Please look at the related [issue](https://github.com/grafana/k6/issues/3857) for more in depth explanation.

Reviewer 2: ```suggestion
For a more in-depth explanation, please refer to the related [issue](https://github.com/grafana/k6/issues/3857).
```

WDYT?

With this version, k6 will start emitting warnings when it detects that in the future, this will break. We recommend using [`import.meta.resolve()`](https://grafana.com/docs/k6/latest/javascript-api/import.meta/resolve/) as a way to make your scripts future proof.

## New features

### Top-level await support [`4007`](https://github.com/grafana/k6/pull/4007)

After the initial native support for ECMAScript modules, k6 can now load those modules asynchronously which also allows `await` to be used in the top-level of a module. That is you can write `await someFunc()` directly in the top most level of a module instead of having to make an async function that you call that can than use `await`.

```javascript
import { open } from 'k6/experimental/fs'
import csv from 'k6/experimental/csv'

let file;
let parser;
(async function () {
	file = await open('data.csv');
	parser = new csv.Parser(file);
})();
```

Can now be written as

```javascript
import { open } from 'k6/experimental/fs'
import csv from 'k6/experimental/csv'

> 

Reviewer 2: ```suggestion
```

When it comes to load testing a website using the k6 browser module, such dynamic requests can result is a high number of similar looking requests which are hard to correlate and find any valuable insights in. This can also lead to tests erroring with a too-many-metrics error due to a high cardinality with metrics that are tagged with similar but dynamically changing URLs.

This is also an issue with synthetic tests, where you may not face the too-many-metrics message, but you will end up with a high number of uncorrelated metric data that can't be tracked over time.

To overcome this issue in the browser module, we have implemented `page.on('metric')` which allows you to define url patterns using regex to match against. When a match is found the `url` and `name` tags for the metric will be replaced by the new `name`.

> #### Example usage

Reviewer 2: Sure, done in cf882b60c467cc251052c0c00f119fea351809ea

      ]
    });
  });

  try {
    // This is only for illustrative purposes, the q query param doesn't affect
    // the website.
    await page.goto('https://test.k6.io/?q=abc123');
    await page.goto('https://test.k6.io/?q=def456');
  } finally {
    await page.close();
  }
}

```

### `ControlOrMeta` support in the keyboard [browser#1457](https://github.com/grafana/xk6-browser/pull/1457)

This will allow us to write the test for all platforms that either work with `Control` or `Meta` when performing keyboard actions. For example `control+click` on windows, and `meta+click` on mac to open a link in a new window.

#### Example usage

<details>
<summary>Expand to see an example usage of `ControlOrMeta`</summary>

```js
  await page.keyboard.down('ControlOrMeta');

  // Open the link in a new tab.
  // Wait for the new page to be created.
  const browserContext = browser.context();
  const [newTab] = await Promise.all([
    browserContext.waitForEvent('page'),
    await page.locator('a[href="/my_messages.php"]').click()
  ]);

  await page.keyboard.up('ControlOrMeta');
```

</details>

## UX improvements and enhancements

> - [browser#1462](https://github.com/grafana/xk6-browser/pull/1462) Enhance `waitForSelector` error message to better reflect why a selector doesn't resolve to an element.

Reviewer: Here (and some other items below) needs the style alignment like we have in our release notes template:

```
_Format as `<number> <present_verb> <object>. <credit>`_:
```

k6 `v0.55.0` is here 🎉! This release includes:

- ⚠️ The deprecated StatsD output has been removed.
- ⚠️ The experimental `k6/experimental/tracing` module has been removed.
- 🆕 URL grouping support in the browser module.
- 🆕 Top-level `await` support.
- 🔐 Complete RSA support for `k6/experimental/webcrypto`.

> ## Breaking changes

Reviewer: @mstoykov should we also list https://github.com/grafana/k6/pull/4009 in breaking changes?

Reviewer 3: I guess so ... I am pretty sure nobody is using this and in practice even if they did - they very likely than gave it to another API that supports all cases

  // Wait for the new page to be created.
  const browserContext = browser.context();
  const [newTab] = await Promise.all([
    browserContext.waitForEvent('page'),
    await page.locator('a[href="/my_messages.php"]').click()
  ]);

  await page.keyboard.up('ControlOrMeta');
```

</details>

## UX improvements and enhancements

- [browser#1462](https://github.com/grafana/xk6-browser/pull/1462) Enhances `waitForSelector` error message to better reflect why a selector doesn't resolve to an element.
- [#4028](https://github.com/grafana/k6/pull/4028) Adds support of SigV4 signing for the `experimental-prometheus-rw` output. This allows users to authenticate with AWS services that require SigV4 signing. Thanks, @obanby for the contribution!
- [#4026](https://github.com/grafana/k6/pull/4026) Allows setting of service.name from the `OTEL_SERVICE_NAME` environment variable for the `experimental-opentelemetry` output. This aligns better with standard OTEL practices. Thanks, @TimotejKovacka for the contribution!
- [browser#1426](https://github.com/grafana/xk6-browser/issues/1426) Instruments `page.waitForTimeout` with tracing which will allow it to be displayed in the timeline.

> ## Bug fixes

Reviewer: It's worth mentioning https://github.com/grafana/k6/pull/4017

### StatsD output removed [`#3849`](https://github.com/grafana/k6/pull/3849)

The StatsD output was deprecated in k6 v0.47.0 and is now removed. You could still output results to StatsD using the community xk6 extension [LeonAdato/xk6-output-statsd](https://github.com/LeonAdato/xk6-output-statsd). Thanks, @LeonAdato for taking over the extension!

### `open` will have a breaking change in the future.

Currently, `open` opens relative files based on an unusual root, similar to how `require` behaved before it was updated for ESM compatibility. To make k6 more consistent, `open` and other functions like it will start handling relative paths in the same way as imports and `require`.
For a more in-depth explanation, please refer to the related [issue](https://github.com/grafana/k6/issues/3857).

With this version, k6 will start emitting warnings when it detects that in the future, this will break. We recommend using [`import.meta.resolve()`](https://grafana.com/docs/k6/latest/javascript-api/import.meta/resolve/) as a way to make your scripts future proof.

> 

Reviewer 3: ### `http.file#data` now truly has the same type as the provided data [`#4009`](https://github.com/grafana/k6/pull/4009)

Previously `http.file#data` was always a slice of byte (`[]byte`) - which was very likely a bug and a leftover from years past.

Both it's documentation, and it is idea though, was to be the original data that was provided when creating it.

k6 `<version>` is here 🎉! This release includes:

- (_optional_) `<highlight of breaking changes>`
- `<Summary of new features>` (_one or multiple bullets_)

## Breaking changes

- Drop `k6/experimental/tracing` a replacement jslib has been released, please look at the [docs on how to transition](https://grafana.com/docs/k6/latest/javascript-api/jslib/http-instrumentation-tempo/#migration-from-k6experimentaltracing), [`#3855`](https://github.com/grafana/k6/pull/3855).

### StatsD output removed [`#3849`](https://github.com/grafana/k6/pull/3849)

> The StatsD output was deprecated in k6 v0.47.0 and is now removed. You could still output results to StatsD using the xk6 extension [LeonAdato/xk6-output-statsd](https://github.com/LeonAdato/xk6-output-statsd). Thanks, @LeonAdato for taking over the extension!

Reviewer 4: * using the _community_ xk6 extension...
